// @wpengine/headless/dist/api/queries/categoryOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.categoryOptions = void 0;
const getCategoryUrlParts = (path) => {
    const parts = Array.isArray(path) ? path : path.split('/').filter(Boolean);
    return {
        basePath: parts.slice(0, 2).join('/'),
        category: parts[1],
        direction: parts[2],
        id: parts[3],
    };
};
const getQueryOptions = (url, postsPerPage) => {
    return {
        variables: {
            first: url.direction === 'after' || !url.direction ? postsPerPage : undefined,
            last: url.direction === 'before' ? postsPerPage : undefined,
            after: url.direction === 'after' ? url.id : undefined,
            before: url.direction === 'before' ? url.id : undefined,
            where: {
                categoryName: url.category,
            },
        },
    };
};
function categoryOptions(urlPath, postsPerPage) {
    const url = getCategoryUrlParts(urlPath !== null && urlPath !== void 0 ? urlPath : []);
    return getQueryOptions(url, postsPerPage);
}
exports.categoryOptions = categoryOptions;
