// @wpengine/headless/dist/api/queries/getPosts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPostsQuery = void 0;
const client_1 = require("@apollo/client");
const LIST_POST_DATA_FRAGMENT_1 = require("./LIST_POST_DATA_FRAGMENT");
const PAGE_INFO_DATA_FRAGMENT_1 = require("./PAGE_INFO_DATA_FRAGMENT");
function getPostsQuery({ fragments } = {}) {
    var _a;
    return client_1.gql `
    ${(_a = fragments === null || fragments === void 0 ? void 0 : fragments.listPostData) !== null && _a !== void 0 ? _a : client_1.gql `
      ${LIST_POST_DATA_FRAGMENT_1.LIST_POST_DATA_FRAGMENT}
    `}
    ${client_1.gql `
      ${PAGE_INFO_DATA_FRAGMENT_1.PAGE_INFO_DATA_FRAGMENT}
    `}
    query GetPosts(
      $where: RootQueryToPostConnectionWhereArgs
      $after: String
      $before: String
      $first: Int
      $last: Int
    ) {
      posts(
        where: $where
        after: $after
        before: $before
        first: $first
        last: $last
      ) {
        pageInfo {
          ...pageInfoData
        }
        nodes {
          ...listPostData
        }
      }
    }
  `;
}
exports.getPostsQuery = getPostsQuery;
